### PR TITLE
Optimizations for GDScript VM

### PR DESCRIPTION
**Pull request by reduz**
* Removed instruction argument count and instruction prefetching. This is now done on the fly. Reduces jumps.
* Removed the address mode switch, addressing is now simple array indexing.
* OPCODE_DISPATCH now goes directly to the next instruction, like in Godot 3.x.

Performance seems to improve significantly (more than 40% faster on release on typed code vs master), here is some quick benchmark I made:

**Master:**
Debug:
Time taken (untyped): 33.708msec.
Time taken (typed): 20.464msec.
Release:
Time taken (untyped): 8.257msec.
Time taken (typed): 5.761msec.

**This PR:**
Debug:
Time taken (untyped): 29.234msec.
Time taken (typed): 16.027msec.
Release:
Time taken (untyped): 6.41msec.
Time taken (typed): 3.311msec.

**Test source code:**

```GDScript
const elems = 65536

func benchmark():

	var arr = []
	for x in range(elems):
		arr.append(randi() % elems)


	var t = Time.get_ticks_usec()

	var acc = 0.0

	for e in arr:
		var e2 = arr[e]
		var e3 = arr[e2]
		acc += e * e2
		acc *= e3 + e
		acc = sqrt(acc)

	
	print("Time taken (untyped): ",(Time.get_ticks_usec() - t)/1000.0,"msec.")		

	var arr2 := PackedInt32Array()
	for x in range(elems):
		arr2.append(randi() % elems)


	t = Time.get_ticks_usec()

	var acc2 : float = 0.0

	for e in arr2:
		var e2 :int = arr2[e]
		var e3 :int = arr2[e2]
		acc2 += e * e2
		acc2 *= e3 + e
		acc2 = sqrt(acc)

	
	print("Time taken (typed): ",(Time.get_ticks_usec() - t)/1000.0,"msec.")		
```
I don't have much else I can use to test performance, so if anyone wants to lend a hand and compare with master (both on debug and release) or test in your production project, it would be very welcome.

### CODE REVIEW

**Comment by Calinou** [comment]
If anyone's curious, I ported the benchmark code to PHP and Python. Benchmark times were measured on an i7-6700K @ 4.4 GHz on Fedora 36.

### PHP

Time taken (PHP 8.1.13 + XDebug): 16.055 msec
Time taken (PHP 8.1.13): 3.538 msec
Time taken (PHP 8.1.13 + [JIT](https://stitcher.io/blog/php-8-jit-setup)): 1.605 msec

<details>
<summary>Code</summary>

```php
<?php

const elems = 65536;

function benchmark() {
	$arr = [];
	for ($i = 0; $i < elems; $i++) {
		$arr[] = mt_rand(0, 2147483647) % elems;
    }

	$t = microtime(true);

	$acc = 0.0;

	foreach ($arr as $e) {
		$e2 = $arr[$e];
		$e3 = $arr[$e2];
		$acc += $e * $e2;
		$acc *= $e3 + $e;
		$acc = sqrt($acc);
    }

	echo "Time taken (PHP): " . (microtime(true) - $t) * 1000 . " msec.";
}

benchmark();
```
</details>

### Python

Time taken (Python 3.9.0): 12.492 msec
Time taken ([Pyston](https://github.com/pyston/pyston) 2.3.4): 8.459 msec

<details>
<summary>Code</summary>

```python
import math
import random
from time import time

elems = 65536

def benchmark():
	arr = []
	for x in range(elems):
		arr.append(random.randint(0, 2147483647) % elems)

	t = time()

	acc = 0.0

	for e in arr:
		e2 = arr[e]
		e3 = arr[e2]
		acc += e * e2
		acc *= e3 + e
		acc = math.sqrt(acc)


	print("Time taken (Python):", (time() - t) * 1000, "msec.")

benchmark()
```
</details>

File: modules/gdscript/gdscript_byte_codegen.h [hunk 1]
```diff
@@ -332,7 +332,12 @@ class GDScriptByteCodeGenerator : public GDScriptCodeGenerator {
 	}
 
 	void append(GDScriptFunction::Opcode p_code, int p_argument_count) {
```
Comment (vnen) [inline modules/gdscript/gdscript_byte_codegen.h:334]: I believe we can remove the argument count here since it's not being used anymore.

**Comment by Byteron** [comment]
> If anyone's curious, I ported the benchmark code to PHP, Python and Ruby. Benchmark times were measured on an i7-6700K @ 4.4 GHz on Fedora 36.
> 
> ## PHP
> * Time taken (PHP 8.1.13 + XDebug): 16.055 msec
> * Time taken (PHP 8.1.13): 3.538 msec
> * Time taken (PHP 8.1.13 + [JIT](https://stitcher.io/blog/php-8-jit-setup)): 1.605 msec[1](#user-content-fn-1-b653e5b2f1b72dcbf21f8402853ecf95)
> 
> Code
> ## Python
> * Time taken (Python 3.9.0): 12.492 msec
> * ~Time taken ([Pyston](https://github.com/pyston/pyston) 2.3.4): 8.459 msec[1](#user-content-fn-1-b653e5b2f1b72dcbf21f8402853ecf95)~
> * Time taken ([Pyston](https://github.com/pyston/pyston) 2.3.5): 6.686 msec[1](#user-content-fn-1-b653e5b2f1b72dcbf21f8402853ecf95)
> 
> Code
> ## Ruby
> * Time taken (Ruby 3.2.0): 7.357 msec
> * Time taken (Ruby 3.2.0 + `--jit`): 4.711 msec[1](#user-content-fn-1-b653e5b2f1b72dcbf21f8402853ecf95)
> 
> Code
> ## Footnotes
> 1. A JIT compiler is not comparable to interpreters in terms of performance. This figure is included for reference only. [↩](#user-content-fnref-1-b653e5b2f1b72dcbf21f8402853ecf95) [↩2](#user-content-fnref-1-2-b653e5b2f1b72dcbf21f8402853ecf95) [↩3](#user-content-fnref-1-3-b653e5b2f1b72dcbf21f8402853ecf95) [↩4](#user-content-fnref-1-4-b653e5b2f1b72dcbf21f8402853ecf95)

Given you are on different Hardware it probably makes sense to also include GDScript Numbers on your 6700K.

**Review by vnen — APPROVED** [review]
Looks good to me, apart from the bit I commented about (which can be addressed later).

**Comment by bruvzg** [comment]
Some tests on M1 mac, using release editor binary (average of 10 runs):

|                          | Time    |
|--------------------------|---------|
| Godot (master, untyped)  | 13.9504 |
| Godot (master, typed)    |  8.9020 |
| Godot (this pr, untyped) | 11.2082 |
| Godot (this pr, typed)   |  6.0292 |
| Python 3.10.9            |  9.1007 |
| Ruby 2.6.10p210          |  8.6488 |

File: modules/gdscript/gdscript_byte_codegen.h (hunk 1 above)
Comment (reduz) [reply modules/gdscript/gdscript_byte_codegen.h:334]: Yeah, I have to, but its used in some places that I was afraid to mess up and introduce some bug, so I will do after its well tested.

**Comment by barathbheeman** [comment]
I'm new to godot, so can someone pls explain why the comparisons are made against python, php and ruby instead of c# and c++ since these are the languages someone might actually use in a godot project?

**Comment by Calinou** [comment]
> I'm new to godot, so can someone pls explain why the comparisons are made against python, php and ruby instead of c# and c++ since these are the languages someone might actually use in a godot project?

C# and C++ are compiled languages, so these aren't fair comparisons to make. It's not possible for an interpreted language to even match a compiled language.

**Comment by jordo** [comment]
> I'm new to godot, so can someone pls explain why the comparisons are made against python, php and ruby instead of c# and c++ since these are the languages someone might actually use in a godot project?

We've been doing a lot of transpiling gdscript to c++ for our game logic in some performance heavy areas... so I have a lot of experience here.  We run a lot of godot gameservers on cloud infrastruction so we pay for every CPU cycle used.  

For GDScript vs C++ we can typically see close to two orders of magnitude of speedup, (usually 40-100X faster), which is why it's not a fair comparison but also illustrates how there can be so much to gain with an AOT language + optimization pass that compiles to machine instructions.

The typed gdscript example code runs in `16.033msec` on my 2019 x86 macbook pro.   

The native c++ version below, executes in `0.021000msec`.  So in this example, actually close to 1000X faster.

```
	int elems = 65536;
	
	std::vector<int> arr = {};
	for (int i = 0; i < elems; i++) {
		arr.push_back(rand() % elems);
	}

	uint64_t t = mach_absolute_time();

	float acc = 0.0;

	for (int e = 0; e < elems; e++) {
		int e2 = arr[e];
		int e3 = arr[e2];
		acc += e * e2;
		acc *= e3 + e;
		acc = sqrt(acc);
	}

	printf("Time taken (typed): %.4f%s",(mach_absolute_time() - t)/1000.0,"msec.");
```

Now we typically don't see a 1000X improvement.  But in this trivial example the native optimization passes that a c++ compiler can do (vectorization, inlining, loop unrolling, et al) seem to really play a factor.  In production we usually see about a 40X-100X performance improvement.

File: modules/gdscript/gdscript_byte_codegen.h (hunk 1 above)
Comment (reduz) [reply modules/gdscript/gdscript_byte_codegen.h]: Alright, done.

**Comment by reduz** [comment]
@jordo We have been discussing this here - https://github.com/godotengine/godot-proposals/issues/6031 and the general consensus seems to be that users would be happy with an optional transpiler to C, but this is obviously Godot 4.0+ material.

**Comment by fire** [comment]
I launched my game and I think this call stack is related.


<details>
  <summary>Stacktrace</summary>
```batch
[Inline Frame] godot.windows.editor.x86_64.llvm.exe!CowData<Variant>::size() Line 133 (c:\Users\ernes\Documents\workspace\godot\core\templates\cowdata.h:133)
[Inline Frame] godot.windows.editor.x86_64.llvm.exe!CowData<Variant>::get(int p_index) Line 155 (c:\Users\ernes\Documents\workspace\godot\core\templates\cowdata.h:155)
[Inline Frame] godot.windows.editor.x86_64.llvm.exe!Vector<Variant>::operator[](int p_index) Line 93 (c:\Users\ernes\Documents\workspace\godot\core\templates\vector.h:93)
godot.windows.editor.x86_64.llvm.exe!Array::operator[](int p_idx) Line 104 (c:\Users\ernes\Documents\workspace\godot\core\variant\array.cpp:104)
godot.windows.editor.x86_64.llvm.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 2872 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript_vm.cpp:2872)
godot.windows.editor.x86_64.llvm.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1830 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript.cpp:1830)
godot.windows.editor.x86_64.llvm.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 711 (c:\Users\ernes\Documents\workspace\godot\core\object\object.cpp:711)
godot.windows.editor.x86_64.llvm.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1159 (c:\Users\ernes\Documents\workspace\godot\core\variant\variant_call.cpp:1159)
godot.windows.editor.x86_64.llvm.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1563 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript_vm.cpp:1563)
godot.windows.editor.x86_64.llvm.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1830 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript.cpp:1830)
godot.windows.editor.x86_64.llvm.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 711 (c:\Users\ernes\Documents\workspace\godot\core\object\object.cpp:711)
godot.windows.editor.x86_64.llvm.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1159 (c:\Users\ernes\Documents\workspace\godot\core\variant\variant_call.cpp:1159)
godot.windows.editor.x86_64.llvm.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1563 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript_vm.cpp:1563)
godot.windows.editor.x86_64.llvm.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1830 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript.cpp:1830)
godot.windows.editor.x86_64.llvm.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 711 (c:\Users\ernes\Documents\workspace\godot\core\object\object.cpp:711)
godot.windows.editor.x86_64.llvm.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1159 (c:\Users\ernes\Documents\workspace\godot\core\variant\variant_call.cpp:1159)
godot.windows.editor.x86_64.llvm.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1563 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript_vm.cpp:1563)
godot.windows.editor.x86_64.llvm.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1830 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript.cpp:1830)
godot.windows.editor.x86_64.llvm.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 711 (c:\Users\ernes\Documents\workspace\godot\core\object\object.cpp:711)
godot.windows.editor.x86_64.llvm.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1159 (c:\Users\ernes\Documents\workspace\godot\core\variant\variant_call.cpp:1159)
godot.windows.editor.x86_64.llvm.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1563 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript_vm.cpp:1563)
godot.windows.editor.x86_64.llvm.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1830 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript.cpp:1830)
godot.windows.editor.x86_64.llvm.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 711 (c:\Users\ernes\Documents\workspace\godot\core\object\object.cpp:711)
godot.windows.editor.x86_64.llvm.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1159 (c:\Users\ernes\Documents\workspace\godot\core\variant\variant_call.cpp:1159)
godot.windows.editor.x86_64.llvm.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1563 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript_vm.cpp:1563)
godot.windows.editor.x86_64.llvm.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1830 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript.cpp:1830)
godot.windows.editor.x86_64.llvm.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 711 (c:\Users\ernes\Documents\workspace\godot\core\object\object.cpp:711)
godot.windows.editor.x86_64.llvm.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1159 (c:\Users\ernes\Documents\workspace\godot\core\variant\variant_call.cpp:1159)
godot.windows.editor.x86_64.llvm.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1563 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript_vm.cpp:1563)
godot.windows.editor.x86_64.llvm.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1830 (c:\Users\ernes\Documents\workspace\godot\modules\gdscript\gdscript.cpp:1830)
[Inline Frame] godot.windows.editor.x86_64.llvm.exe!Node::_gdvirtual__ready_call() Line 241 (c:\Users\ernes\Documents\workspace\godot\scene\main\node.h:241)
godot.windows.editor.x86_64.llvm.exe!Node::_notification(int p_notification) Line 154 (c:\Users\ernes\Documents\workspace\godot\scene\main\node.cpp:154)
godot.windows.editor.x86_64.llvm.exe!Object::notification(int p_notification, bool p_reversed) Line 792 (c:\Users\ernes\Documents\workspace\godot\core\object\object.cpp:792)
godot.windows.editor.x86_64.llvm.exe!Node::_propagate_ready() Line 188 (c:\Users\ernes\Documents\workspace\godot\scene\main\node.cpp:188)
godot.windows.editor.x86_64.llvm.exe!Node::_propagate_ready() Line 178 (c:\Users\ernes\Documents\workspace\godot\scene\main\node.cpp:178)
godot.windows.editor.x86_64.llvm.exe!Node::_set_tree(SceneTree * p_tree) Line 2577 (c:\Users\ernes\Documents\workspace\godot\scene\main\node.cpp:2577)
godot.windows.editor.x86_64.llvm.exe!SceneTree::initialize() Line 411 (c:\Users\ernes\Documents\workspace\godot\scene\main\scene_tree.cpp:411)
godot.windows.editor.x86_64.llvm.exe!OS_Windows::run() Line 1273 (c:\Users\ernes\Documents\workspace\godot\platform\windows\os_windows.cpp:1273)
godot.windows.editor.x86_64.llvm.exe!widechar_main(int argc, wchar_t * * argv) Line 181 (c:\Users\ernes\Documents\workspace\godot\platform\windows\godot_windows.cpp:181)
[Inline Frame] godot.windows.editor.x86_64.llvm.exe!_main() Line 203 (c:\Users\ernes\Documents\workspace\godot\platform\windows\godot_windows.cpp:203)
godot.windows.editor.x86_64.llvm.exe!main(int argc, char * * argv) Line 226 (c:\Users\ernes\Documents\workspace\godot\platform\windows\godot_windows.cpp:226)
godot.windows.editor.x86_64.llvm.exe!WinMainCRTStartup() (Unknown Source:0)
godot.windows.editor.x86_64.llvm.exe!WinMainCRTStartup() (Unknown Source:0)
kernel32.dll!00007ff933b426bd() (Unknown Source:0)
ntdll.dll!00007ff93502dfb8() (Unknown Source:0)
```
</details>

**Comment by TokisanGames** [comment]
I ran our game w/ nearly 30k lines of code and a thousand assets and compared beta 10. I didn't expect much. Though we have a lot of code we're not bound by gdscript or the CPU. We're bound by the renderer and the GPU. I rarely see any scripts with significant performance issues on the monitors, and fix them if I do.

Testing the PR build, I do have a lot more renderer errors and visual artifacts (prob due to other PRs). I did not get any gdscript errors. 

| What | B10 | This PR |
|-----|-------|--------|
| Loading the project from the manager | 24.7s | 38s (1st) 17.6 (2nd) |
| Loading the game to title screen | 14.1s | 27.4s (1st) 13.6 (2nd) |
| Loading the first level | 8.4s | 16.6s (1st) 8.5s (2nd) |
| Playing, running around one area | 100-145fps | 100-145fps |


Beta 10
image

PR
image

**Comment by Zireael07** [comment]
Why the big difference between 1st and 2nd run?

**Comment by Norrox** [comment]
> I'm new to godot, so can someone pls explain why the comparisons are made against python, php and ruby instead of c# and c++ since these are the languages someone might actually use in a godot project?
What Calinou said and we can already use C# with Godot so why compare the two ? :D

**Comment by akien-mga** [comment]
Note that the way binaries are compiled has a big impact on performance comparisons. Official beta builds are a lot more optimized than CI builds, so to do an accurate comparison using the PR's CI build, I would suggest downloading a CI build from the `master` branch around the time that PR was made (e.g. https://github.com/godotengine/godot/actions/runs/3818020117).

**Comment by cridenour** [comment]
Ran my project with around ~8K lines of statically typed GDScript in both debug and release. I don't have any benchmark numbers at the moment, but the procedural world generation felt faster - and felt _much_ faster on MacOS. 

No errors anywhere in the project.

**Comment by LinuxUserGD** [comment]
With Godot 4 beta 10 I noticed that string concatenation seems to be 700 times slower compared to Python:
GDScript (16.63s):
```gdscript
func string() -> int:
	var x : String = ""
	for i in range(0, 300000):
		x += " "
	return x.length()
```

Python (0.023s):
```python
def string():
    x = ""
    for i in range(0, 300000):
        x += " "
    return len(x)
```

**Comment by reduz** [comment]
@LinuxUserGD most likely Python is optimized for string processing (because that's a common use case for it) and their string class keeps track of all the additions to later merge them when you read from it. Godot is not designed with string processing in mind hence this is slow. We have a StringBuilder class we use on the C++ side that could be exposed to script if there is really demand, but probably there isn't.

**Comment by holzmb** [comment]
will this get backported to godot 3

**Comment by AThousandShips** [comment]
@holzmb GDScript has been completely reworked for 4.x

**Comment by reduz** [comment]
@holzmb
This optimization is already in Godot 3.x:
> * OPCODE_DISPATCH now goes directly to the next instruction, **like in Godot 3.x.**
What is not there is the addressing mode optimization, but that won't happen because its too much work.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**